core/src/main/java/tc/oc/pgm/loot/LootableMatchModule.java

```diff
@@ -0,0 +1,302 @@
+                    .collect(Collectors.joining(", ")));
+
+    // Find all Caches that the opened inventory is part of
+    final List<Fillable> fillables = new ArrayList<>();
+    for (Cache cache : caches) {
+      if (filterPredicate.test(cache.region()) && filterPredicate.test(cache.filter())) {
+        fillables.add(new FillableCache(cache));
+      }
+    }
+    // If the inventory is not in any Cache, just fill it directly
+    if (fillables.isEmpty()) {
+      fillables.add(new FillableInventory(inventory));
+    }
+
+    fillables.forEach(fillable -> fillable.fill(opener, fillers));
+  }
+
+  abstract class Fillable {
+
+    abstract Stream<Inventory> inventories();
+
+    void fill(MatchPlayer opener, List<FillerDefinition> fillers) {
+      // Build a short list of Fillers that are NOT cooling down from a previous fill
+      final List<FillerDefinition> coolFillers =
+          fillers.stream()
+              .filter(
+                  filler ->
+                      filledAt.putUnlessNewer(new Pair<>(this, filler), filler.getRefillInterval())
+                          == null)
+              .collect(Collectors.toList());
+
+      // Find all the Inventories for this Fillable, and build a map of Fillers to the subset
+      // of Inventories that they are allowed to fill, based on the filter of each Filler.
+      // Note how duplicate inventories are skipped.
+      final SetMultimap<FillerDefinition, Inventory> fillerInventories = HashMultimap.create();
+      inventories()
+          .distinct()
+          .forEach(
+              inventory -> {
+                final Predicate<Filter> passes = filterPredicate(inventory.getHolder());
+                if (passes == null) return;
+                for (FillerDefinition filler : coolFillers) {
+                  if (passes.test(filler.getFilter())) {
+                    fillerInventories.put(filler, inventory);
+                  }
+                }
+              });
+
+      fillerInventories
+          .asMap()
+          .forEach(
+              (filler, inventories) -> {
+                if (filler.cleanBeforeRefill()) {
+                  inventories().forEach(Inventory::clear);
+                }
+              });
+
+      final Random random = new Random();
+
+      fillerInventories
+          .asMap()
+          .forEach(
+              (filler, inventories) -> {
+                // For each Filler, build a mutable list of slots that it can fill.
+                // (27 is the standard size for single chests)
+                final List<InventorySlot> slots = new ArrayList<>(inventories.size() * 27);
+                inventories.forEach(
+                    inv -> {
+                      for (int index = 0; index < inv.getSize(); index++) {
+                        if (inv.getItem(index) == null) {
+                          slots.add(new InventorySlot(index, inv));
+                        }
+                      }
+                    });
```

Comment:
```suggestion
                for (Inventory inv : inventories) {
                  for (int index = 0; index < inv.getSize(); index++) {
                    if (inv.getItem(index) == null) 
                      slots.add(new InventorySlot(index, inv));
                  }
                }
```
Sometimes lambdas aren't any shorter, i'd use a traditional loop for this inner-most loop